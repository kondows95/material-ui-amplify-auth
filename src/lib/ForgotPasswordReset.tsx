import React, { FormEvent, ChangeEvent } from 'react';
import { Box, Container, TextField } from '@material-ui/core';
import { useIntl } from 'react-intl';
import FormTitle from './FormTitle';
import { MyButton } from 'material-ui-basic-parts';

type Props = {
    authState: string | null;
    forgotPasswordSubmit: (email: string, confirmationCode: string, password: string) => void;
    loading: boolean;
    error: string;
    email: string;
};

const ForgotPasswordReset: React.FC<Props> = (props) => {
    const { formatMessage } = useIntl();

    type Form = { [key: string]: string };
    const [form, setForm] = React.useState<Form>({ email: '', password: '', confirmationCode: '' });

    type FormKey = 'email' | 'password' | 'confirmationCode';
    const handleChangeValue = (fieldName: FormKey) => (event: ChangeEvent<HTMLInputElement>): void => {
        const newForm = { ...form };
        newForm[fieldName] = event.target.value;
        setForm(newForm);
    };

    const handleSubmit = (event: FormEvent): void => {
        event.preventDefault();
        props.forgotPasswordSubmit(props.email, form['confirmationCode'], form['password']);
    };

    const msgId = 'MUAA.ForgotPasswordReset';
    const msgIdTitle = msgId + '.Title';
    const msgIdCode = msgId + '.Code';
    const msgIdPassword = msgId + '.Password';
    const msgIdSubmit = msgId + '.Submit';
    const content = (
        <Container data-testid={msgId} component="main" maxWidth="xs">
            <form onSubmit={handleSubmit}>
                <Box display="flex" flexDirection="column" mt={8}>
                    <FormTitle
                        data-testid={msgIdTitle}
                        label={formatMessage({ id: msgIdTitle, defaultMessage: 'Please confirm your email' })}
                    />
                    <Box display="flex" justifyContent="center" fontWeight={600} color="error.main">
                        {props.error}
                    </Box>
                    <Box width="100%" my={2}>
                        <TextField
                            label={formatMessage({ id: msgIdCode, defaultMessage: 'Confirmation Code' })}
                            onChange={handleChangeValue('confirmationCode')}
                            value={form.confirmationCode}
                            variant="outlined"
                            required
                            fullWidth
                            inputProps={{ 'data-testid': msgIdCode }}
                        />
                    </Box>

                    <Box width="100%" my={2}>
                        <TextField
                            label={formatMessage({ id: msgIdPassword, defaultMessage: 'New Password' })}
                            type="password"
                            autoComplete="new-password"
                            onChange={handleChangeValue('password')}
                            value={form.password}
                            variant="outlined"
                            required
                            fullWidth
                            inputProps={{ 'data-testid': msgIdPassword }}
                        />
                    </Box>

                    <Box width="100%" mt={4} mb={2} className="relative">
                        <MyButton
                            type="submit"
                            label={formatMessage({ id: msgIdSubmit, defaultMessage: 'Submit' })}
                            color="primary"
                            loading={props.loading}
                            data-testid={msgIdSubmit}
                        />
                    </Box>
                </Box>
            </form>
        </Container>
    );

    return props.authState === 'forgotPasswordReset' ? content : null;
};

export default ForgotPasswordReset;
