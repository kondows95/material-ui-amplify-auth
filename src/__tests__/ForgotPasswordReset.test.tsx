import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import TestLib from 'react-testing-library-wrapper';
import { IntlProvider } from 'react-intl';
import ForgotPasswordReset from '../lib/ForgotPasswordReset';

let tLib: TestLib;
let forgotPasswordSubmit: jest.Mock;

beforeEach(() => {
    forgotPasswordSubmit = jest.fn();
});

describe('Basic Usage', () => {
    const msgs = {
        'MUAA.ForgotPasswordReset.Title': '#MUAA.ForgotPasswordReset.Title',
        'MUAA.ForgotPasswordReset.Code': '#MUAA.ForgotPasswordReset.Code',
        'MUAA.ForgotPasswordReset.Password': '#MUAA.ForgotPasswordReset.Password',
        'MUAA.ForgotPasswordReset.Submit': '#MUAA.ForgotPasswordReset.Submit',
    };
    beforeEach(() => {
        tLib = new TestLib(
            (
                <IntlProvider locale="en" messages={msgs}>
                    <ForgotPasswordReset
                        authState={'forgotPasswordReset'}
                        forgotPasswordSubmit={forgotPasswordSubmit}
                        loading={false}
                        error={''}
                        email={'test@example.com'}
                    />
                </IntlProvider>
            )
        );
    });

    it('Click submit button', () => {
        tLib.changeValue('MUAA.ForgotPasswordReset.Code', '12345');
        tLib.changeValue('MUAA.ForgotPasswordReset.Password', 'myPassword');
        tLib.click('MUAA.ForgotPasswordReset.Submit');
        expect(forgotPasswordSubmit).toHaveBeenCalledWith('test@example.com', '12345', 'myPassword');
    });

    it('Snapshot', () => {
        expect(tLib.render.asFragment()).toMatchSnapshot();
    });
});

it('Defalt Intl labels', () => {
    tLib = new TestLib(
        (
            <IntlProvider locale="en">
                <ForgotPasswordReset
                    authState={'forgotPasswordReset'}
                    forgotPasswordSubmit={forgotPasswordSubmit}
                    loading={false}
                    error={''}
                    email={'test@example.com'}
                />
            </IntlProvider>
        )
    );
    //For <TextField> default label.
    const html = tLib.render.container.innerHTML;
    expect(html).toContain('Confirmation Code');
    expect(html).toContain('New Password');
    //For textContent.
    expect(tLib.get('MUAA.ForgotPasswordReset.Title').textContent).toBe('Please confirm your email');
    expect(tLib.get('MUAA.ForgotPasswordReset.Submit').textContent).toBe('Submit');
});

it('If authState is not forgotPasswordReset, return empty.', () => {
    tLib = new TestLib(
        (
            <IntlProvider locale="en">
                <ForgotPasswordReset
                    authState={'forgotPassword'}
                    forgotPasswordSubmit={forgotPasswordSubmit}
                    loading={false}
                    error={''}
                    email={'test@example.com'}
                />
            </IntlProvider>
        )
    );
    expect(tLib.render.container.textContent).toBe('');
});
