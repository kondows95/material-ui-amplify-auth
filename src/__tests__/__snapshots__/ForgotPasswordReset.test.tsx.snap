// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Usage Snapshot 1`] = `
<DocumentFragment>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs"
    data-testid="MUAA.ForgotPasswordReset"
  >
    <form>
      <div
        class="MuiBox-root MuiBox-root-454"
      >
        <div
          class="MuiBox-root MuiBox-root-455"
          data-testid="MUAA.ForgotPasswordReset.Title"
        >
          <div
            class="MuiBox-root MuiBox-root-456"
          >
            <div
              class="MuiBox-root MuiBox-root-457"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
                />
              </svg>
            </div>
          </div>
          <div
            class="MuiBox-root MuiBox-root-467"
          >
            #MUAA.ForgotPasswordReset.Title
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-468"
        >
          MyError
        </div>
        <div
          class="MuiBox-root MuiBox-root-469"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink="false"
            >
              #MUAA.ForgotPasswordReset.Code
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                data-testid="MUAA.ForgotPasswordReset.Code"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-528 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-530"
                >
                  <span>
                    #MUAA.ForgotPasswordReset.Code *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-532"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink="false"
              for="standard-adornment-password"
            >
              #MUAA.ForgotPasswordReset.Password
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
              required=""
            >
              <input
                aria-invalid="false"
                autocomplete="current-password"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                data-testid="MUAA.ForgotPasswordReset.Password"
                id="standard-adornment-password"
                required=""
                type="password"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
              >
                <button
                  aria-label="toggle password visibility"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
                  data-testid="MUAA.ForgotPasswordReset.Password.Visible"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-528 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-530"
                >
                  <span>
                    #MUAA.ForgotPasswordReset.Password
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-552 relative"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
            data-testid="MUAA.ForgotPasswordReset.Submit"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              #MUAA.ForgotPasswordReset.Submit
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </form>
  </main>
</DocumentFragment>
`;
